common/v2/features/ProtectTransaction/withProtectStateFactory.tsx

```diff
@@ -0,0 +1,321 @@
+import { useCallback, useContext, useEffect, useRef } from 'react';
+
+import {
+  CryptoScamDBInfoResponse,
+  CryptoScamDBNoInfoResponse
+} from 'v2/services/ApiService/CryptoScamDB/types';
+import { TUseStateReducerFactory } from 'v2/utils';
+import CryptoScamDBService from 'v2/services/ApiService/CryptoScamDB/CryptoScamDB';
+import { Asset, ITxReceipt, Network, NetworkId, WalletId } from 'v2/types';
+import { GetBalanceResponse, GetLastTxResponse } from 'v2/services/ApiService/Etherscan/types';
+import { EtherscanService } from 'v2/services/ApiService/Etherscan';
+import { getNetworkById, NetworkContext } from 'v2/services/Store/Network';
+import { AssetContext, getAssetByUUID } from 'v2/services/Store/Asset';
+import useMediaQuery from 'v2/vendor/react-use/useMediaQuery';
+import { BREAK_POINTS } from 'v2/theme';
+import { WALLETS_CONFIG } from 'v2/config';
+
+import { SendFormCallbackType } from './types';
+
+export interface WithProtectState {
+  stepIndex: number;
+  protectTxShown: boolean;
+  protectTxEnabled: boolean;
+  cryptoScamAddressReport: CryptoScamDBNoInfoResponse | CryptoScamDBInfoResponse | null;
+  etherscanBalanceReport: GetBalanceResponse | null;
+  etherscanLastTxReport: GetLastTxResponse | null;
+  mainComponentDisabled: boolean;
+  receiverAddress: string | null;
+  network: Network | null;
+  asset: Asset | null;
+  isWeb3Wallet: boolean;
+  web3WalletName: string | null;
+}
+
+export interface WithProtectApiFactory {
+  withProtectState: WithProtectState;
+  formCallback: SendFormCallbackType;
+
+  handleTransactionReport(receiverAddress?: string): Promise<void>;
+  setMainTransactionFormCallback(callback: SendFormCallbackType): void;
+  goOnNextStep(): void;
+  goOnInitialStepOrFetchReport(receiverAddress?: string): void;
+  showHideTransactionProtection(showOrHide: boolean): void;
+  setReceiverInfo(receiverAddress: string, network: NetworkId | null): Promise<void>;
+  setProtectionTxTimeoutFunction(cb: (txReceiptCb?: (txReciept: ITxReceipt) => void) => void): void;
+  invokeProtectionTxTimeoutFunction(cb: (txReceipt: ITxReceipt) => void): void;
+  clearProtectionTxTimeoutFunction(): void;
+  setWeb3Wallet(isWeb3Wallet: boolean, walletTypeId?: WalletId | null): void;
+}
+
+export const WithProtectInitialState: Partial<WithProtectState> = {
+  stepIndex: 0,
+  protectTxShown: false,
+  protectTxEnabled: false,
+  receiverAddress: null,
+  network: null,
+  cryptoScamAddressReport: null,
+  etherscanBalanceReport: null,
+  etherscanLastTxReport: null,
+  asset: null,
+  isWeb3Wallet: false,
+  web3WalletName: null
+};
+
+const numOfSteps = 3;
+
+const WithProtectConfigFactory: TUseStateReducerFactory<
```

Comment:
In case you want to migrate, we've started using `useReducer` with a classic `[state, dispatch]`, Flux Standard Action (FSA) payloads and actionCreators. Not sure which middleware to adopt but for the moment it works very well when combined within a custom hook (to serve the actionCreators). I makes the state changes more readable and with `reinspect` we even got access to ReduxDevTools.